### PR TITLE
Hook up error listeners for Chat and Calling adapters

### PR DESCRIPTION
**Pull request by allenplusplus**
# What
Hook up error listeners for Chat and Calling adapters.

# Why
It should address this bug where HTTP error screen is shown when token/thread is incorrect: https://skype.visualstudio.com/SPOOL/_workitems/edit/2457313.

# How Tested
Tested via unit test and manually check composites on storybook.

# Process & policy checklist

- [X] I have updated the project documentation to reflect my changes if necessary.
- [X] I have read the [CONTRIBUTING](https://github.com/Azure/communication-ui-library/blob/main/CONTRIBUTING.md) documentation.

**Is this a breaking change?**

- [ ] This change causes current functionality to break.

### CODE REVIEW

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts
Comment (JamesBurnside) [inline packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts:72]: We should really avoid doing this as it adds large amounts of arbitrary time to unit tests. Whats the waiting for? Is there a way we can use jest mock timers to speed up triggering the wait?

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:211]: nit: this error should indicate this is a joining error, `Error joining call - user is already in the call`.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:282]: Out of interest, how is the context error used currently? Do we make use of it or is it in the state for a user to query? I'm worried that because we call setError on every error that an error a user wants to act on will be overwritten before they get the chance to?

Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:282]: Right, that is a valid concern. One way to not lose the error is to have Contoso register the error listener and they will be able to receive all errors without missing any. @PorterNan , is the original design for state to hold the error so it can be retrieved? If so what do you think of James's concern about the error being overwritten with a new error?

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts:72]: It is because the dispose() runs some asynchronous code but dispose doesn't return a promise to await on. The waitWithBreakCondition is a way to wait so the asynchronous code in dispose() completes before we validate it in the test. In practice this is almost immediate since it is mocked out but still asynchronous. The waitWithBreakCondition waits for 100ms at a time in a loop. On my PC it only waits once before we can continue on with the test verification. We can shorten this 100ms even more if we want to do that. Another thing we can do is make adapter's dispose() return a promise but this changes the public API so will be a breaking change.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Sounds good, updated to be `Error joining call - user is already in the call`

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts
Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts:72]: If the adapter dispose is itself firing off awaitable async functions that we are not awaiting we should definitely update this to be a `Promise<void>` regardless of the breaking change

Comment (JamesBurnside) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts:72]: Taking a look presumably this will simply become:
```
  public dispose(): Promise<void> {
    this.callClient.offStateChange(this.onClientStateChange);
    await this.callAgent.dispose();
  }
```

Comment (allenplusplus) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts:72]: Right, that looks to be the change we'll need to make it awaitable.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (smsohan) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:430]: Should we extract a generic error handling method to DRY things up? It seems the pattern is, we want to handle errors by setting it to the context and then emitting an event. In some cases, we want to return a default value after that. Thinking something like this:

```TypeScript
withErrorHandling(async () => {
  await this.handlers.onParticipantRemove(userId);
})

// When we have a default value

withErrorHandling(async () => {
  return this.deviceManager.getMicrophones();
},  Promise.resolve([] as AudioDeviceInfo[]))
```

**Comment by allenplusplus** [comment]
Add breaking-change label. The API file has couple changes to make the of the adapter's methods into arrow functions but this shouldn't break people although it looks different in the API file. The breaking-change is changing CallAdapter's dispose function to return a Promise<void> which may break people depending on their usage of this function.

File: packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts]: Updated this so that CallAdapter's dispose function returns a Promise. The ChatAdapter's dispose function is unchanged (still return void) since there is nothing to await. Also removed the waitWithBreakCondition from both tests.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Thanks Sohan, I think that's a good idea, I've updated this PR with the withErrorHandling wrapper function.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (JamesBurnside) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:485]: out of interest, whats the different between these syntaxes?

**Review by JamesBurnside — APPROVED** [review]

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Approving so this comment isn't blocking - but still wondering if we'll run into a problem with errors stomping over each other

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:485]: The second style creates an arrow function with the current 'this' variable enclosed. The first style does not. This means if the first style is passed to another function and called, the 'this' variable refers to that other function's 'this' and not the original 'this' of the original adapter.

For example:

`<AnotherComponent getState={adapter.getState} />`

If getState is in first style, calling it inside AnotherComponent will return undefined because 'this.state' refers to AnotherComponent's 'this'. But in the second style the above should work since the 'this' is refering to the adapter's 'this' instead of the AnotherComponent's 'this'.

Comment (JamesBurnside) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:485]: Ah I just naively assumed class methods would retain this as the class instance! Thanks!

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Synced up with Nan and James offline. Looks like initially we are planning to have one error in the context and this is shown in the error bar. For the one error case, it is fine to just show the latest error. In the future the design might change to allow for some history of errors but for now I'll consider this resolved as it seems to be the current design but please let me know if there is any concerns.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:485]: Is the change to arrow functions necessary for the main thrust of this PR (i.e., hooking up error handling?). Since both changes are both functionally and in SLOC, I'd request separate PRs for the two (land the arrow function first if it's needed for error handling).

OTOH, I'd like to understand the need better.
MDN [discourages the use of arrow functions for object methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions): 
image

While MDN doesn't have anything to say for classes, this [blog post nicely explains](https://javascriptweblog.wordpress.com/2015/11/02/of-classes-and-arrow-functions-a-cautionary-tale/) why it's better to not use arrow functions in classes either: Arrow functions get added as porperties, so (i) they don't show up on the object's prototype chain; (ii) a new copy of the function is created for each instance of the class.

Now, (ii) is unlikely to cause performance issues here, since we don't expect to have many instances of the Adapter. Still, this is an anti-pattern so we should have a really good reason to use arrow functions here, and should add very obvious documentation and caution in the code so that folks aren't surprised down the line.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:177]: Consider adding private methods to `AzureCommunicationCallAdapter` and `AzureCommunicationChatAdapter` to further redue boilerplate:

```
return this.handleAsyncError(
    async () => {
        return await this.deviceManager.getMicrophones();
    },
    []
)
```

Possibly adding `handleSyncError` as well if necessary.

Changes recommended:
* Naming: `withErrorHandling` sounds like a decorator to me. This is just a function call.
* Add a private method that can pull out `this.context` and `this.emitter` to reduce boilerplate.
* Make the generic method(s) smarter so it can infer the return type.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Let's rename the state field to be `lastError` to better capture the semantics.
We should also provide a `clearLastError()` method in the stateful client to allow Contoso (or the error bar `onClose` handler) to clear the error.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts:71]: I'd recommend splitting this into many small tests, each for a single error generated. In that case, I'd also make the mock stateful call client creation local. Something like:

```

const errorString = 'some arbitrary reusable error string';

describe('AzureCommunicationCallAdapter handles errors correctly', () => {
    test('when dispose fails', async () => {
      // Setup
      const statefulCallClient = createMockStatefulCallClient();
     statefulCallClient.dispose = (): void => {
        raise Error(errorString );
     }
     const [adaptor, errorListener] = newCallAdaptorWithErrorListener(statefulCallClient);  // Helper function in this test file to reduce boilerplate.
     // Act
     await adapter.dispose();
     // Assert
     expect(adapter.getState().error?.message).toBe(errorString );
     expect(errorListener.capturedError).toBe(errorString );
 
    }
}
```

Now, this test doesn't have any of the boilerplate that isn't necessary to test *only the `dispose` method*.
And, this test fails iff there's a bug in dispose. You don't have to go hunting through logs on test failure to figure out what's broken.

Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts:71]: This increases the total number of lines by a bit, but it makes each test a lot smaller and targeted.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (prprabhu-ms) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: The rename doesn't have to be in this PR.

**Comment by prprabhu-ms** [comment]
Sorry for the late intercept here. Might be worth resolving the discussion on approach we want to take: https://github.com/Azure/communication-ui-library/pull/482

This PR is partially in line with what I'm proposing there, but a bunch of the implementation bits will change. I want to shift most of the heavy-lifting of capturing errors to the stateful client instead of the adapter.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Going to leave that for another PR as it seems to be a feature add.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Thanks for the links Prathmesh. The arrow functions are not necessary for this PR and I've removed them from this PR. We may want to decide on which style to use and how users should use the API as we have used different styles in different places.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts]: Thanks for the feedback, I've refactored this as per your requirements, please take a look if it works for you.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts
Comment (allenplusplus) [reply packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts]: Thanks, I've updated the test as per these requirements, please take a look if it works for you.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts:15]: Use `camelCase` for constants.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.test.ts:73]: Thanks!

These tests look much better now. Each test is easy to understand.

File: packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts
Comment (prprabhu-ms) [inline packages/react-composites/src/composites/CallComposite/adapter/AzureCommunicationCallAdapter.ts:492]: +1 great documentation.

File: packages/communication-react/review/communication-react.api.md
Comment (prprabhu-ms) [inline packages/communication-react/review/communication-react.api.md:141]: Should all these API changes be gone now?

**Review by prprabhu-ms — COMMENTED** [review]
Thanks for the quick turnaround.

As I said, lets hold off more work until we decide on the direction.

**Comment by prprabhu-ms** [comment]
Let's close this PR in favor of https://github.com/Azure/communication-ui-library/pull/510

**Comment by prprabhu-ms** [comment]
Clsoing this to avoid confusion. Can always reopen ;)

**Pull request closed without merging**
